Handle primary and replica clients in iothreads

minchopaskal opened the pull request:
# Problem

Primary and replica clients are still handled inside main thread due to data race and synchronization issues. This PR solves this issue by making it possible for them to be handled in IO threads with the hope it would potentially also increase performance.

# Overview

## Current flow

Since it's outside the scope of the review we won't go into too much detail.

### RDB replication

When a replica node connects to a primary (f.e via the REPLICAOF command) the first thing to do is replicate the RDB from primary -> replica. The primary creates a replica client which starts the process by sending a PING, following more handshake messages. Most of them are REPLCONF commands from replica -> primary which set up how the RDB will be replicated (see `syncWithMaster`). We have three degrees of freedom here:
* full or partial sync - depending on whether the replica already has some of the data.
* with or without dual channel RDB replication - default is with - replica opens a second channel for receiving RDB data which offloads keeping the incomming replication data from the primary to replica (see #13732)
* on-disk or diskless - either save the RDB on disk before loading it on the replica or directly read it from socket.
After RDB replication is done the replica node creates a master client which receives replication stream.

### Replication stream

Whenever an event that changes the dataset on the primary occurs it is replicated to all of its replicas(f.e see `propagateNow`). That happens by writing to the replication buffer which is a list of nodes. This buffer is shared amongst replicas (for more detail see #9166). This list is append-only and replicas only read forward. The nodes are refcounted so whenever a replica reads from a node it holds a refcount to it and when it's done with it it decreases the refcount and moves to the next one (see `_writeToClientSlave`).
Since this list can grow indefinitely we need to trim it at some point. Trimming happens in `incrementalTrimReplicationBacklog`. General idea is we always look at the front of the list, if the refcount of the node is 0 we know all of the replicas have moved forward so we don't need it and it's safe to be trimmed. If the next node still has a refcount of 0 we can also trim it and so on. Note here that we never trim everything we could since it's possible there are too much nodes to free and we could freeze the server, so that's why it's done incrementally.

### ACK and WAIT

Replica nodes periodically send `REPLCONF ACK` messages to primary to acknowledge that they are alive and the replication offset they've reached.
Moreover, a master can receive a WAIT command which generally waits for a number of replicas to reach the current replication offset. The primary client sends a `GETACK` command to the replicas so it receives their `ACK` messages ASAP.

## Moving the clients to IO threads

Since clients first participate in a handshake and an RDB replication phases it was decided they are moved to IO-thread after RDB replication is done. For primary client this was trivial as the master client is created only after RDB sync (+ some additional checks one can see in `isClientMustHandledByMainThread`). Replica clients though are moved to IO threads immediately after connection (as are all clients) so currently in `unstable` replication happens while this client is in IO-thread. In this PR it was moved to main thread after receiving the first `REPLCONFG` message from the replica, but it is a bit hacky and we can remove it. I didn't find issues between the two versions.

## Primary client (replica POV)

We have few issues here:
- during `serverCron` a `replicationCron` is ran which periodically sends `REPLCONF ACK` message to the master, also checks for timed-out master. In order to prevent data races with master the behaviour of replication cron with run_with_preiod was replication to IO-thread. It periodically sends the primary client to main thread and during processing it in `processClientsFromIOThread` it checks if a replication cron is due and runs it if so. This is the same behaviour as the `IOThreadClientsCron` with the exception it's not ran on every `IOThreadCron` cycle. There is the detail here that if the primary client is the only client on the replica instance and it resides in IO-thread it will inadvertantly be send to main thread each time `IOThreadClientsCron` is ran. That's why the primary client was excluded from there. Moreover, in order to run the `IOThreadReplicationCron` we cached pointer to the master client inside the `IOThread` structure in order not to search for it in the clients list everytime the cron is called. That necessitates the pointer is nullified each time the client moves to main thread. Thus even if master is disconnected and reconnected/replicaof is changed we'll always have a valid master pointer. 

- data races with main thread - specifically `lastinteraction` and `read_reploff` members of the primary client that are written to in `readQueryFromClient` could be accessed at the same time from main thread during execution of `INFO REPLICATION`(`genRedisInfoString`). To solve this the members were duplicated so if the client is in IO-thread it writes to the duplicates and they are synced with the original variables each time the client is send to main thread ( that means `INFO REPLICATION` could potentially return stale values)

- During freeClient the primaryClient is fetched to main thread but when caching it the thread id will remain the id of the IO thread it was from. We change it back to the main thread id, in case the master is resurrected, so it remains in main thread during the rdb replication process.

- During `exitScriptTimedoutMode` the master is queued for reprocessing (i.e process any pending commands while it was blocked ASAP). We do that by putting it in a unblocked_clients list, which are processed in the next `beforeSleep` cycle. Since this will create a contention between main and IO thread, we just skip this queueing in `unblocked_clients` and just queue the client to main thread - the `processClientsFromIOThread` will process the pending commands just as main would have.

## Replica client (primary POV)

We move the client after RDB replication is done and after replication backlog is fed with its first message.
We do that so that the client's reference to the first repl-backlog node is initialized before it's read from IO-thread, and not contended with it.

### Shared replication buffer

On that note - currently in unstable the replication buffer is shared amongst clients. This is done via clients holding references to the nodes inside the buffer. A node from the buffer can be freed once each replica client has read it and send its contents. The reference is `client->ref_repl_buf_node`. The replication buffer is written to by main thread by `feedReplicationBuffer` and the refcounting is intrusive - it's inside the replication-buffer nodes themselves.
Since the replica client changes the refcount (decreases the refcount of the node it has just read, and increases the refcount of the next node it starts to read) we have a data race with main thread. Moreover, main thread also updates the used size of the node - how much it has written to it, compared to its capacity which the replica client relies on to know how much to read. Obviously replica being in IO-thread creates another data race here. To mitigate these issues a few new variables(names are subject to change) were added to the client's struct:
- `ref_repl_start_node` - starting node this replica is reading from inside IO-thread
- `ref_last_node` - last node in the repl buffer this replica has last seen
- `ref_last_used` - last seen used value of the last seen last node (hope that's clear!)
These values are only allowed to be updated in main thread. The client keeps track of how much it has read into the buffer via the old `ref_repl_buf_node`. Generally while in IO-thread the replica client will now keep refcount of the `ref_repl_start_node` until it's processed all the nodes up to `ref_last_node` - at that point its returned to main thread which can safely update the refcounts.
The last node reference is there so the replica knows when to stop reading from the repl buffer - imagine that replica reads from the last node of the replication buffer while main thread feeds data to it - we will create a data race on the `used` value (`_writeToClientSlave`(IO-thread) vs `feedReplicationBuffer`(main)). That's why this value is updated just before the replica is being send to IO thread.

Example:

* Replication buffer state at time N:
   | node 0| ... | node M, used_size K |
* replica caches `ref_repl_start_node`=0, `ref_last_node`=M and `ref_last_used`=K
* replica moves to IO thread and processes all the data it sees
* Replication buffer state at time N + 1:
   | node 0| ... | node M, used_size Full | |node M + 1| |node M + 2, used_size L|, where Full > M
* replica moves to main thread at time N + 1, at this point following happens
   - refcount to node 0 is decreased
   - `ref_repl_start_node` becomes node M (we still have Full-M bytes to process from there)
   - refcount to node M is increased (now all nodes from 0 up to M-1 including can be trimmed unless some other replica holds reference to them)
   - And just before the replica is send back to IO thread the following are updated:
   - `ref_last_node` ref becomes node M+2
   - `ref_last_used` becomes L

Note that replica client is only moved to main if it has processed all the data it knows about (i.e up to `ref_last_node` + `ref_last_used`)

### Replica clients kept in main as much as possible

During implementation an issue arose - how fast is the replica client able to get knowledge about new data from the replication buffer and how fast can it trim it. In order for that to happen ASAP whenever a replica is moved to main it remains there until the replication buffer is fed new data.  At that point its put in the pending write queue and special cased in handleClientsWithPendingWrites so that its send to IO thread ASAP to write the new data to replica. Also since each time the replica writes it whole repl data it knows about, it's send to main then `processClientsFromIOThread` is able to immediately update the refcounts and trim whatever it can.

### ACK messages from primary

Slave clients need to periodically read REPLCONFG ACK messages from client. Since replica can remain in main thread indefinitely if no DB change occurs, a new variable `io_last_ack_time` was added to the clients struct so that it can be checked against - if some time has passed since last ACK read we move back the replica to IO-thread so it has a chance to read it. Another case is when the whole repl data is written the client may stay in IO thread if some specified time has passed since the last ACK read..

`io_last_ack_time` actually replicates the `repl_ack_time` but is kept in millisecond granularity. Also, there may be contention with main thread during `genRedisInfoString` on that variable, so whenever IO thread needs to write to it it just updates `io_last_ack_time` and the value of `repl_ack_time` is synced with it when the replica is moved to main (just like `lastinteraction` and `read_reploff` for the primary client).

### Replicas during shutdown

During shutdown the main thread pauses write actions and periodically checks if all replicas have reached the same replication offset as the primary node. During `finishShutdown` that may or may not be the case. Either way a client data may be read from the replicas and even we may try to write any pending data to them inside `flushSlavesOutputBuffers`. In order to prevent races all the replicas from IO threads are moved to main via `fetchClientFromIOThread`. A cancel of the shutdown should be ok, since the mechanism employed by `handleClientsWithPendingWrites` should return the client back to IO thread when needed (not sure how to test that though and I'll need some help on this one.)

## Notes

Some tests were added, and a few timing issues (mainly with the TSAN job) were fixed.

minchopaskal committed:
Handle MASTER client in IO threads

minchopaskal committed:
remove redundant set to REPL_STATE_CONNECTED

minchopaskal committed:
Move slave clients to io threads

minchopaskal committed:
remove debug prints

minchopaskal committed:
Fix consistency-check tests

minchopaskal committed:
Fix consistency-check tests again

minchopaskal committed:
WIP

minchopaskal committed:
checkmarx test

minchopaskal committed:
Atomic handling of replBufBlock::refcount.

Handle replBufBlock::used data-race by caching it before sending client
to io-thread

minchopaskal committed:
remove unnecessary barriers

minchopaskal committed:
fix replica not assigned to io-thread

minchopaskal committed:
Fix master client sending ack in io-thread

minchopaskal committed:
Keep replica in main thread as much as possible

minchopaskal committed:
Keep client in main thread while waiting for new repl data

minchopaskal committed:
WIP FIX keep replica in main thread, while waiting for repl data

minchopaskal committed:
fix replica client not reading ack, fix master client sending too much ack

minchopaskal committed:
Remove wrong assert - replica may stay in io-thread on partial write and be send to main from ACK query

minchopaskal committed:
revert replBufBlock::refcount to be non-atomic

minchopaskal committed:
Fix slave assigned to main thread not updating its ref_repl_start_node

minchopaskal committed:
fix replBufBlock refcount update in processClientsFromIOThreads

minchopaskal committed:
refactoring

minchopaskal committed:
remove assert

minchopaskal committed:
more refactoring

minchopaskal committed:
revert disabled master/slave in iothreads

fix spelling

minchopaskal committed:
handle iothread slaves waiting indefinitely in main

minchopaskal committed:
fix missing void param

minchopaskal committed:
fix leftover data race issues

minchopaskal committed:
fix repl_ack_time bug

minchopaskal committed:
fix replica not sending enough ACK messages

minchopaskal committed:
fix IO-master being put in unblocked_clients list on script timeout

remove io_repl_cron flag

minchopaskal committed:
add continue-on-error for tsan daily jobs

minchopaskal committed:
incr wait time for waiting on resumed replicas

minchopaskal committed:
Add tests

Lower ack read time to 500 so we don't slow down slave client ACK reads

minchopaskal committed:
a bit of refactoring

minchopaskal committed:
Fix replication-iothreads test

minchopaskal committed:
try more frequent io replication cron

minchopaskal committed:
Fix master client still cached in iothread after fetchClientFromIOThread

minchopaskal committed:
disable iothreads tests for valgrind daily job

remove some AI slop from replication-iothreads tests

minchopaskal committed:
incr wait timeout for replication-iothreads tests

minchopaskal committed:
add wait_for_condition for psync2 check for equality of replica and subreplica vs master's digest

minchopaskal committed:
Update master clients data each time its send from io to main thread

minchopaskal committed:
disable fail-fast for sanitizer-thread matrix in daily ci

minchopaskal committed:
Fix clients from iothread data update. Disable iothread tests for valgrind daily.

minchopaskal committed:
Disable repl tests for io-threads job in daily

minchopaskal committed:
fix valgrind-no-malloc-usable-size-test job run call

minchopaskal committed:
Fix leftover timing issues

minchopaskal committed:
fix replication-iothreads tests

skaslev commented on src/iothread.c:
```diff
@@ -139,17 +204,48 @@ void fetchClientFromIOThread(client *c) {
         return 1;
     }
+
+    /* If RDB replication is done it's save to move the master client to an IO thread */
```
`save` -> `safe`

skaslev commented on src/iothread.c:
```diff
@@ -139,17 +204,48 @@ void fetchClientFromIOThread(client *c) {
+    if (c->flags & CLIENT_MASTER &&
+        server.repl_state == REPL_STATE_CONNECTED &&
+        server.repl_rdb_ch_state == REPL_RDB_CH_STATE_NONE)
+        return 0;
```
Braces would make it easier to read because of the big multi-line condition in the `if ()`
```
    {
        return 0;
    }
```

sggeorgiev commented on src/server.h:
```diff
@@ -3106,6 +3152,7 @@ void freeReplicaReferencedReplBuffer(client *replica);
 void replicationFeedMonitors(client *c, list *monitors, int dictid, robj **argv, int argc);
 void updateSlavesWaitingBgsave(int bgsaveerr, int type);
 void replicationCron(void);
+void replicationSendAck(void);
```
Do you really need to expose this function? It looks like you use it only in replication.c.

minchopaskal replied:
ah yes it was exposed before, but reverted later. Will revert that too.

minchopaskal committed:
Fix timing issues with tsan tests

small refactoring

sggeorgiev commented on src/networking.c:
```diff
@@ -304,9 +311,20 @@ static inline int _prepareClientToWrite(client *c) {
+     * since it checks how far they've read into the replication buffer and may
+     * return true without them already being in the pending write queue. That
+     * happens when main thread has written into the replication buffer while
+     * they were in IO thread. Later handleClientsWithPendingReplies will deal
```
handleClientsWithPendingReplies does not exist

sggeorgiev commented on src/iothread.c:
```diff
@@ -59,22 +87,54 @@ void enqueuePendingClientsToMainThread(client *c, int unbind) {
+    }
+    if (c->flags & CLIENT_SLAVE) {
+        c->ref_last_node = listLast(server.repl_buffer_blocks);
+        c->ref_last_node_used = ((replBufBlock*)listNodeValue(c->ref_last_node))->used;
```
Do we need to increase `refcount` of `replBufBlock` here. What will happens if the block is trimmed.

minchopaskal replied:
The block can't be trimmed here as we are in main thread and main thread is the only one allowed to trim the repl buffer.
`ref_last_node` is just a cache to the last node in the replication buffer this client had seen when it was sent from main to IO thread. The actual "reference" is to the `ref_start_node` which may or may not be equal to the last node, but we can assert that `ref_start_node` is always previous or equal to `ref_last_node`. Since `ref_start_node` hold a reference to a node that is previous to `ref_last_node` then both nodes can't have been trimmed yet.

For IO threads trimming happens after a client is returned to main thread and the refcounts of the nodes are updated.

sggeorgiev commented on src/networking.c:
```diff
@@ -2115,23 +2141,55 @@ static inline int _writeToClientNonSlave(client *c, ssize_t *nwritten) {
+        /* Main thread can safely write to repl buffer nodes so we handle
+         * refcounting here. For IO thread replicas refcount is handled in
+         * processClientsFromIOThread */
+        if (c->running_tid == IOTHREAD_MAIN_THREAD_ID) {
```
Is it possible this if to be true if all writes are executed in iothreads?

minchopaskal replied:
yes, this is the code-path for replica clients in main thread. If we run redis with `--io-threads 1`(default btw) we will go through here.

minchopaskal committed:
fix tests ::tsan does not exist for cluster tests

minchopaskal committed:
flush output buffers of slaves in io-threads

add tsan option for sentinel tests

minchopaskal committed:
Fix flushSlavesOutputBuffers for IO-thread slaves and address review comments

minchopaskal committed:
Fix read_reploff after prefetch changes

minchopaskal committed:
Fix Tsan issues and some issues related to rebase on prefetch changes

skaslev commented on src/server.c:
```diff
@@ -1778,6 +1778,23 @@ static void sendGetackToReplicas(void) {
+
+    if (server.io_threads_num == 1) return;
+
+    /* Since we want to read the ACK ASAP we make sure to return any replicas
```
Can we generalize (and rename) `putSlavesNeedingAckReadInPendingClientsToIOThreads()` with a `flags` argument (say whether to also check `slaveFromIOThreadNeedsAckRead()` before sending to io threads) and use it here too?

minchopaskal committed:
remove --config io-threads 2 option for sentinel tests in daily tsan job

minchopaskal committed:
Renaming

minchopaskal committed:
Remove resetClientInternal from tsan.sup

minchopaskal committed:
Remove redundant IOThreadReplicationCron

minchopaskal committed:
Only send replica to io-thread if have repl data or we *know* there is pending ACK

minchopaskal committed:
cosmetic change

Co-authored-by: Yuan Wang <wangyuancode@163.com>

minchopaskal committed:
Refactor after review comments

Removed IOThreadReplicationCron

Unify logic of sending replica client back to IO thread in one place in
beforeSleep

Small fixes and code refactors

minchopaskal committed:
Small refactoring after review

minchopaskal committed:
fix spelling

minchopaskal committed:
Fix tsan errors

minchopaskal committed:
Fix review comments, fix code commets

minchopaskal committed:
Check if replica is scheduled for close before sending it to IO thread

ShooterIT committed:
simplify

minchopaskal committed:
small refactor

minchopaskal committed:
update comments

minchopaskal committed:
add check

minchopaskal committed:
fix running_tid in fetchClientFromIOThread

minchopaskal committed:
fix comments

minchopaskal committed:
zero-initialize pending_read

minchopaskal committed:
Add check for ref_repl_buf_node when updating repl node refs for io-thread

minchopaskal committed:
Update src/networking.c

Co-authored-by: Yuan Wang <wangyuancode@163.com>

minchopaskal committed:
Update src/iothread.c

Co-authored-by: Yuan Wang <wangyuancode@163.com>

minchopaskal committed:
more code cleanup

minchopaskal committed:
review fixes

minchopaskal committed:
remove unneeded check
